Visualization/bokeh_holoviews_datashader.ipynb

```diff
@@ -0,0 +1,560 @@
+    "def update_histogram(bounds=bounds):\n",
+    "    \n",
+    "    selection = (ds.data['ra'] > bounds[0]) & \\\n",
+    "                (ds.data['dec'] > bounds[1]) & \\\n",
+    "                (ds.data['ra'] < bounds[2]) & \\\n",
+    "                (ds.data['dec'] < bounds[3])\n",
+    "    \n",
+    "    selected_mag = ds.data.loc[selection]['phot_g_mean_mag']\n",
+    "    \n",
+    "    frequencies, edges = np.histogram(selected_mag)\n",
+    "    \n",
+    "    hist = hv.Histogram((np.log(frequencies), edges))\n",
+    "    return hist"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%output size=150\n",
+    "dmap = hv.DynamicMap(update_histogram, streams=[box])\n",
+    "datashade(points, cmap=bokeh.palettes.Viridis256) * box_plot + dmap"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Notice that when you select different regions of the left panel with the box select tool, the histogram on the right is updated."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 3: Images\n",
+    "\n",
+    "The next example demonstrates image visualization at the pixel level with datashader."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# First get a sensor image\n",
+    "import lsst.daf.persistence as dafPersist\n",
+    "\n",
+    "butler = dafPersist.Butler('/project/shared/data/Twinkles_subset/output_data_v2')\n",
+    "subset = butler.subset('calexp')\n",
+    "dataid = subset.cache[0]\n",
+    "image = butler.get('calexp', immediate=True, dataId=dataid)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%opts Image  [height=600 width=650]\n",
+    "%%opts Bounds (color='white')\n",
+    "#%%output size=200\n",
+    "\n",
+    "# Make a fake image just for testing purpose\n",
+    "#zz = np.random.poisson(100, size=(4000, 4000))\n",
+    "#xx, yy = np.meshgrid(np.arange(4000), np.arange(4000))\n",
+    "#zz += xx\n",
+    "#bounds=(0, 0, 4000, 4000)   # Coordinate system: (left, bottom, top, right)\n",
+    "#img = hv.Image(zz, bounds=bounds).options(colorbar=True, cmap=bokeh.palettes.Viridis256, logz=True)\n",
+    "\n",
+    "# Use an actual sensor image\n",
+    "bounds_img = (0, 0, image.getDimensions()[0], image.getDimensions()[1])\n",
+    "img = hv.Image(np.log10(image.image.array), bounds=bounds_img).options(colorbar=True, cmap=bokeh.palettes.Viridis256) # Another option is logz=True\n",
+    "\n",
+    "boundsxy = (0, 0, 0, 0)\n",
+    "box = streams.BoundsXY(source=img, bounds=boundsxy)\n",
+    "bounds = hv.DynamicMap(lambda bounds: hv.Bounds(bounds), streams=[box])\n",
+    "\n",
+    "rasterize(img) * bounds"
+   ]
```

Comment:
I get a RuntimeWarning; might want to suppress warnings:
```
RuntimeWarning: invalid value encountered in log10
```

Reply:
I also see really interesting patterning in the downsampled image (see attached). Might be interesting to comment on this (is it real background fluctuations?). What algorithm is datashader using to create the zoomed out image? Is it downsampled, average, median, etc.?

![image](https://user-images.githubusercontent.com/5605123/45909697-0719af00-bdc9-11e8-969e-b00414f914e9.png)

Reply:
I noticed those large angular scale gradations in the background level as well. I haven't been able to find the algorithm used by datashader for downsampling in a few minutes of poking around.